### ISSUE TITLE
People registration

### ISSUE THREAD

**Issue by oliverbarnes** [BlueRidgeLabs/kimball#5]
As a Person (research subject)
I need to register 
So I'm available for research interviews 

Acceptance criteria / technical details

- [ ] validation of email and phone number (phone number validated through twilio)
- [ ] add uniqueness constraints for email and phone number
- [ ] track validation of email and phone number in DB

### Notes for implementation:

* route the public registration controller under a namespace

### Notes for future milestones:

- (make sure no non-validated numbers are called or emails sent to non-validated addresses)
- (somehow set it up so that the new features run in parallel to the old ones, e.g. by routing them under a namespace)

## Update

Requirements discussed on [WIP PR](https://github.com/BlueRidgeLabs/kimball/pull/16):

(&OR == And or, meaning they can enter both or one, but not none. anything not market as optional is required (name, postal code etc.))

```
#  first_name                       :string(255)
#  last_name                        :string(255)
&OR phone number #  email_address                    :string(255)
optional #  address_1                        :string(255)
optional #  address_2                        :string(255)
optional #  city                             :string(255)
optional #  state                            :string(255)
#  postal_code                      :string(255)
&OR email #  phone_number                     :string(255) 
choice between email or phone #  preferred_contact_method         :string(255)
```

this bit below is about giving the ability to select what kind of device they have, as well as how the person connects to the internet. Potentially it should have been another model, but still works as an attribute on person. 
```
#  primary_device_id                :integer
#  primary_device_description       :string(255)
optional #  secondary_device_id              :integer
optional #  secondary_device_description     :string(255)
#  primary_connection_id            :integer
#  primary_connection_description   :string(255)
optional #  secondary_connection_id          :integer
optional #  secondary_connection_description :string(255)
```

here's the code for connection ID and mapping:
```
# in /config/initializers/device_mappings.rb
Logan::Application.config.device_mappings = {
  desktop: 0, laptop: 1, smartphone: 2, tablet: 3
}

Logan::Application.config.connection_mappings = {
  phone: 0, home_broadband: 1, other: 2, public_computer: 3, public_wifi: 4\
}

```

**Comment by oliverbarnes** [comment]
@marcoow @cromulus I forget - what is meant by 'track validation of email and phone number in DB'?

**Comment by oliverbarnes** [comment]
also, does 'phone number validated through twilio' mean verification that the phone number is real? 

should we leave any twilio stuff for milestone 3, which focuses on SMS communication? I'd say the quickest we get to the search milestone, the best

**Comment by marcoow** [comment]
>  I forget - what is meant by 'track validation of email and phone number in DB'?

Just a boolean flag in the DB that says whether email/phone number are validated or not

> also, does 'phone number validated through twilio' mean verification that the phone number is real?

exactly

> should we leave any twilio stuff for milestone 3, which focuses on SMS communication? I'd say the quicker we get to the search milestone, the better

makes sense; is that ok with you @cromulus?

**Comment by oliverbarnes** [comment]
Cool, same goes for email then, validation that it's real? Would that be done by email confirmation? Maybe we should call it verification and have a `verified_*` boolean field for each. 

@cromulus if you're MIA I'll go ahead and implement email confirmation and add the `verified_email` flag to `email_address`, leaving phone verification through SMS for milestone 3, and then get started with search.

**Comment by oliverbarnes** [comment]
come to think of it, email verification won't be trivial here since we're not using devise, in which case it'd be as simple as adding the `:confirmable` option to the devise user model.

given that unverified emails will simply mean some requests for interviews will bounce in the case the emails don't exist, is it safe to assume it's not a priority and we can add verification later?

I'm going ahead and merging the PR to start on search, and adding a separate issue for contact info verification for milestone 3.

**Comment by marcoow** [comment]
@oliverbarnes: we can implement our own email validation - just addin a new verification_token attribute with a random token.

**Comment by oliverbarnes** [comment]
how do you mean? how would the email be verified automatically without confirmation by the person?

**Comment by marcoow** [comment]
I mean we just implement email verification ourselves - storing a token in the db, sending that to the email address and then have a route that verifies the token and marks the email address as verified.

**Comment by oliverbarnes** [comment]
ah, ok I see. sure, we can definitely roll our own, creating a mailer to send the verification token, etc. I didn't mean it's hard, just that it'll take some hours rather than minutes with devise. As far as I understand it, this is not critical to this milestone.

It will be critical once we get to scheduling invitations, since there's the case where a person inserts their contact info incorrectly by mistake, and they'll unreachable for invites. We can tackle the email verification then (milestone 3), and SMS verifications on milestone 4?

I might be missing something, of course :) let me know

**Comment by marcoow** [comment]
Yeah that's what I meant. Wouldn't really recommend adding devise given its huge size. Makes sense for me moving it to later milestones though.

**Comment by oliverbarnes** [comment]
agreed, also doesn't make sense to use devise if the person won't be signing in.
